### PR TITLE
Implement a new Team Billing where Owners can conveniently manage a paid plan for their Team

### PR DESCRIPTION
**Pull request by jankeromnes**
## Description

Allow teams to easily purchase and manage a Team Plan directly from their team's Settings.

TODO:
- [x] Implement a new Team Plan shape that's directly linked to / owned by a team: https://github.com/gitpod-io/gitpod/pull/9655
- [x] Allow subscribing via Chargebee directly from your team Settings
- [x] Create a "synthetic" user subscription when you're a member of a paid Team Plan
- [x] Transparently update the Team Plan quantity in Chargebee when users join or leave a team
- [x] Pro-rate charge for new members joining in the middle of a billing cycle
- [x] Allow managing your Billing in team Settings via a "Team Chargebee Portal"
- [x] Align the new UI with #8352 

## Related Issue(s)

Fixes https://github.com/gitpod-io/gitpod/issues/7759

## How to test

1. Create a Team
2. Invite other team members
3. Go to Team Settings > Plans
3. Buy a Team Plan, verify billing amount matches member count, verify all members are indeed upgraded
4. Add/remove members, verify that the Team Plan billing amount is updated accordingly
5. Verify that cancelling, re-upgrading, deleting the team all work as expected

## Release Notes

```release-note
Implement a new Team Billing where Team Owners can conveniently manage a paid plan for their Team
```

## Documentation

- [ ] /werft without-vm=true
- [ ] /werft with-helm=true
- [x] /werft with-payment=true
- [x] /werft with-clean-slate-deployment

### CODE REVIEW

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To complete the [pull request process](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process), please ask for approval from **jankeromnes** after the PR has been reviewed.

Associated issue: *#7759*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

<details open>
Needs approval from an approver in each of these files:

- **[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/8041#issuecomment-1030154588" title="Approved">jankeromnes</a>*

Associated issue: *#7759*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by roboquat** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/gitpod-io/gitpod/pull/8041#issuecomment-1030154588" title="Approved">jankeromnes</a>*

Associated issue: *#7759*

The full list of commands accepted by this bot can be found [here](https://prow.gitpod-dev.com/command-help?repo=gitpod-io%2Fgitpod).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[components/dashboard/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/dashboard/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-db/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-db/OWNERS)~~ [jankeromnes]
- ~~[components/gitpod-protocol/OWNERS](https://github.com/gitpod-io/gitpod/blob/main/components/gitpod-protocol/OWNERS)~~ [jankeromnes]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by mads-hartmann** [comment]
@jankeromnes just a heads up that we're no longer using the `/approved` command (or the associated label) so this PR will need reviews from the CODEOWNERS once it's out of draft before it can be merged ☺️

**Comment by jankeromnes** [comment]
Agree on this getting a tiny bit stale, but still actively worked on and almost finished! Will get finished once higher priorities are out of the way.

**Comment by jankeromnes** [comment]
Currently blocked on https://github.com/gitpod-io/gitpod/issues/8829

**Comment by jankeromnes** [comment]
Me: Hello?
Dark hallway echo: _...hello? ...hello? ...hello? ..._

/werft run

**Comment by jankeromnes** [comment]
Made some free disk space, let's try this again:

/werft run

**Comment by jankeromnes** [comment]
### State of PR

- Feature complete (purchase, cancel, invite/join/leave/delete team, pro-rated invoices, members get unlimited hours)
- Initial functional UI in Team Settings

### Screenshots

<table><tr><td>
Screenshot 2022-05-05 at 18 28 05
</td><td>
Screenshot 2022-05-05 at 18 28 56
</td><td>
Screenshot 2022-05-05 at 18 29 08
</td><td>
Screenshot 2022-05-05 at 18 29 47
</td></tr></table>

### Caveats

- [ ] Depends on https://github.com/gitpod-io/gitpod/pull/9655 (needs to be rebased post-merge)
- [ ] Has a temporary "DEBUG" commit to assist with testing
- [ ] Exact plan details are still boilerplate (to be fixed soon)
- [ ] We have the option to hide the "Team Billing" menu item in Team Settings in order to soft-launch this
- [ ] Should test that legacy team plans still work (should not be affected by the new code)

File: components/ee/payment-endpoint/src/chargebee/upgrade-helper.ts
Comment (jankeromnes) [inline components/ee/payment-endpoint/src/chargebee/upgrade-helper.ts:67]: FIXME: `remainingMs` (actual time left) can be longer than `billingPeriodMs` (average duration of a month) -- should make this more accurate, or prevent it from going over 1 month:

```
currentTermRemainingRatio=1.0184325058939843, diffInCents=2342"
```

(i.e. more than one month remaining, invoicing more than a full month of 23€)

**Review by jldec — APPROVED** [review]
Great stuff - everything works (small pro-rata comment)
Hold before merging and complete caveats.

File: components/ee/payment-endpoint/src/chargebee/upgrade-helper.ts
Comment (jankeromnes) [reply components/ee/payment-endpoint/src/chargebee/upgrade-helper.ts]: Fixed by getting the accurate total duration of the current billing cycle (some months are longer, some shorter, pro-rate should be based on actual cycle duration).

**Comment by jldec** [comment]
@jankeromnes please hold this PR while we review the question of seat limits.

- **Old team-plans** : Invite URLs are limited to the number of seats purchased and will fail beyond that limit.
- **New teams** : Invite URLs can be shared by any member and do not (currently) have seat limits.

Some options to mitigate this: (open to other ideas)
 
- Simple toggle on the billing page to allow / disallow additional members to join
- Text (number) field on the billing page to limit member count
- Limit invite URLs to single use, and only owners can create

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (AlexTugarev) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1459]: why `""`?

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (AlexTugarev) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1435]: is it worth adding a check for equality?

the whole process is a chain of many asyncs/promises, so it's possible for additions/removals to balance out.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (AlexTugarev) [inline components/server/ee/src/workspace/gitpod-server-impl.ts:1206]: 😢  
at least this is not used to id the customers. 
just the regular note: `User.getPrimaryEmail` is not stable.

File: components/ee/payment-endpoint/src/chargebee/team-subscription-handler.ts
Comment (AlexTugarev) [inline components/ee/payment-endpoint/src/chargebee/team-subscription-handler.ts:114]: does this still apply?

**Review by AlexTugarev — COMMENTED** [review]
Left a few comments. Looks very good!

**Comment by jldec** [comment]
@jankeromnes - regarding the PR hold to resolve the question of seat limits:
Let's please merge the PR after making a change to hide the "Billing" menu in Team settings (or using a feature flag).
We are leaning toward option 3 to limit who can generate member invite URLs to just team owners.  
cc: @geropl

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1434]: Sure, I'll add a check and error log.

File: components/server/ee/src/workspace/gitpod-server-impl.ts
Comment (jankeromnes) [reply components/server/ee/src/workspace/gitpod-server-impl.ts:1455]: @AlexTugarev Would you like this to be changed/improved somehow?

**Review by AlexTugarev — APPROVED** [review]

**Comment by jankeromnes** [comment]
This error is still here 😭

```
QueryFailedError: ER_PARSE_ERROR: You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near 'WHERE deleted = true' at line 1
    at QueryFailedError.TypeORMError [as constructor] (/app/node_modules/typeorm/error/TypeORMError.js:9:28)
    at new QueryFailedError (/app/node_modules/typeorm/error/QueryFailedError.js:13:28)
    at Query.<anonymous> (/app/node_modules/typeorm/driver/mysql/MysqlQueryRunner.js:222:57)
    at Query.<anonymous> (/app/node_modules/mysql/lib/Connection.js:526:10)
    at Query._callback (/app/node_modules/mysql/lib/Connection.js:488:16)
    at Query.Sequence.end (/app/node_modules/mysql/lib/protocol/sequences/Sequence.js:83:24)
    at Query.ErrorPacket (/app/node_modules/mysql/lib/protocol/sequences/Query.js:92:8)
    at Protocol._parsePacket (/app/node_modules/mysql/lib/protocol/Protocol.js:291:23)
    at Parser._parsePacket (/app/node_modules/mysql/lib/protocol/Parser.js:433:10)
    at Parser.write (/app/node_modules/mysql/lib/protocol/Parser.js:43:10)
    at Protocol.write (/app/node_modules/mysql/lib/protocol/Protocol.js:38:16)
    at Socket.<anonymous> (/app/node_modules/mysql/lib/Connection.js:88:28)
    at Socket.<anonymous> (/app/node_modules/mysql/lib/Connection.js:526:10)
    at Socket.emit (node:events:390:28)
    at Socket.emit (node:domain:475:12)
    at addChunk (node:internal/streams/readable:315:12)
```

**Comment by jankeromnes** [comment]
Many thanks @AlexTugarev for spotting and fixing the SQL error above in https://github.com/gitpod-io/gitpod/pull/9957 ! 🎉 ❤️

This is now ready to go then. 🚀

/unhold

**Pull request merged**
